### PR TITLE
Rewrite specifically for Sum and Prod to remove Join

### LINKED ISSUE #59: Optimize `Sum`s of `MakeVector`s and `Join`s
### Please describe the purpose of filing this issue

Not a drastic improvement by any means, but something we can keep in mind:

```
reduce(at.concatenate(*tensors)) -> reduce(reduce(tensor) for tensor in tensors)
```
*Ignoring any axis complexities*

```python
import pytensor
import pytensor.tensor as pt
import numpy as np

x = pt.vector("x")
y = pt.vector("y")

f1 = pytensor.function([x, y], pt.sum(pt.concatenate((x, y))))
f2 = pytensor.function([x, y], pt.sum((pt.sum(x), pt.sum(y))))
f3 = pytensor.function([x, y], pt.add(pt.sum(x), pt.sum(y)))

pytensor.dprint(f1)
print()
pytensor.dprint(f2)
print()
pytensor.dprint(f3)

x_val = np.random.rand(100_000)
y_val = np.random.rand(200_000)

%timeit f1(x_val, y_val)
%timeit f2(x_val, y_val)
%timeit f3(x_val, y_val)
```
```
Sum{acc_dtype=float64} [id A] ''   1
 |Join [id B] ''   0
   |TensorConstant{0} [id C]
   |x [id D]
   |y [id E]

Sum{acc_dtype=float64} [id A] ''   3
 |MakeVector{dtype='float64'} [id B] ''   2
   |Sum{acc_dtype=float64} [id C] ''   1
   | |x [id D]
   |Sum{acc_dtype=float64} [id E] ''   0
     |y [id F]

Elemwise{Add}[(0, 0)] [id A] ''   2
 |Sum{acc_dtype=float64} [id B] ''   1
 | |x [id C]
 |Sum{acc_dtype=float64} [id D] ''   0
   |y [id E]
544 µs ± 27.5 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
270 µs ± 5.11 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
270 µs ± 8.86 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

### PR DESCRIPTION
**Pull request by HangenYuu**
## Description

Rewrite specifically for Sum and Prod to remove Join. So graph that starts as
```bash
Sum{axes=None} [id A] 1
 └─ Join [id B] 0
    ├─ 0 [id C]
    ├─ x [id D]
    └─ y [id E]
```
will become
```bash
Add [id A] 2
 ├─ Sum{axes=None} [id B] 1
 │  └─ x [id C]
 └─ Sum{axes=None} [id D] 0
    └─ y [id E]
```

## Related Issue

- [ ] Closes #59
- [ ] Related to #

## Checklist

- [x] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [ ] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)

## Type of change

- [x] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):

### CODE REVIEW

**Comment by ricardoV94** [comment]
This may overlap/be redundant with https://github.com/pymc-devs/pytensor/pull/888

**Review by HangenYuu — COMMENTED** [review]
My rewrite depends on #346 to get to the final value for `Sum` as the implementation introduces a `MakeVector`. For it to work for Prod, I need to add a similar one for `Prod`.

File: pytensor/tensor/rewriting/math.py [hunk 1]
```diff
@@ -1754,6 +1755,30 @@ def local_reduce_broadcastable(fgraph, node):
                 # -- in this case we can remove the reduction completely
                 return [new_reduced.astype(odtype)]
 
+@register_canonicalize
+@register_uncanonicalize
+@register_specialize
+@node_rewriter([Sum, Prod])
```
Comment (ricardoV94) [inline pytensor/tensor/rewriting/math.py:1761]: When there's nothing special about Sum and Prod we should apply the rewrites to all CAReduce operations, which also include stuff like Max, All, Any, ...

See the related PR I linked to in the comment

Comment (HangenYuu) [reply pytensor/tensor/rewriting/math.py:1761]: Oops, 2nd time. I did not see this one show up in #59 😄.

**Comment by ricardoV94** [comment]
> My rewrite depends on #346 to get to the final value for `Sum` as the implementation introduces a `MakeVector`. For it to work for Prod, I need to add a similar one for `Prod`.

#346 was also myopic in treating sum of make vectors more special than any CAReduce of makevector

File: pytensor/tensor/rewriting/math.py (hunk 1 above)
Comment (ricardoV94) [reply pytensor/tensor/rewriting/math.py:1761]: No they are not the same exactly.

There's a rewrite for reduction along axis0 for a sum along axis 0.

My PR extends this to any axis.

There's then the question of multiple axis, of which axis=None is the most extreme (all axes). This PR can cover that case. 

We may also want to think about multiple but not all axis. In what cases can we reduce first and join later?

**Comment by HangenYuu** [comment]
> This may overlap/be redundant with https://github.com/pymc-devs/pytensor/pull/888

**Pull request closed without merging**